File: src/shared/develop/capture-view.ts
/**
 * The capture's files as a VIEWER lists them (R6 of
 * `docs/capture-renditions.md`, 2026-09-21): the words on the lightbox's
 * chips, what each one can show, and what the `Develop` verb under the
 * picture carries away.
 *
 * The lightbox writes NOTHING — *"la visionneuse n'écrit aucun choix, c'est
 * juste de la visualisation"*. Looking at the camera's JPEG beside the proxy
 * is looking; the document is exactly as it was when the sheet closes. What
 * carries the choice is the VERB: pressing `Develop` while the ARW is on
 * screen opens the develop on that rendition (`viewedRendition`).
 *
 * Pure and DOM-free: the fetching, the slicing of a RAW's render and the
 * object URLs are `use-capture-view.ts`'s.
 */

import { isRawImage } from '../library/assets';
import type { CaptureInput, PixelSize, Rendition } from '../media/renditions';
import { openingRendition } from '../media/renditions';
import type { WinnowAssetRow } from '../sources/winnow/client';

/**
 * The rows a viewer can put on screen: everything but the SENSOR's. A viewer
 * draws files; developing a sensor plane is the Develop tool's work, and a
 * chip that can only say "open Develop" is a control that disappoints.
 */
export function viewableRenditions(rows: readonly Rendition[]): Rendition[] {
  return rows.filter((r) => r.role !== 'sensor');
}

/** The chip's word — the fidelity chip's own: `Proxy`, else the file's name. */
export function viewLabel(row: Rendition): string {
  return row.role === 'proxy' ? 'Proxy' : row.name;
}

/**
 * The line under the picture while that file is shown: what it is, its
 * pixels where measured, its weight, and whether a click will fetch it.
 */
export function viewFacts(row: Rendition, formatBytes: (n: number) => string): string {
  const parts: string[] = [];
  if (row.role === 'proxy') parts.push('the proxy, where the picture opens');
  else if (row.reach === 'embedded') parts.push('the render the camera wrote inside the RAW');
  else parts.push('the file itself');
  if (row.pixels) parts.push(`${row.pixels.width} × ${row.pixels.height}`);
  if (row.bytes != null) parts.push(formatBytes(row.bytes));
  if (!row.here && row.assetId) parts.push('fetched from its instance on request, held for this session');
  return parts.join(' · ');
}

/**
 * What the `Develop` verb hands to the tool: the rendition being viewed, as
 * the roll stores it — null for the one the picture opens on, which is what
 * the workbench stores for "chose nothing", so a look at the proxy never
 * writes a choice the pill would then draw as one.
 */
export function viewedRendition(rows: readonly Rendition[], viewing: string | null): string | null {
  if (!viewing) return null;
  const row = rows.find((r) => r.id === viewing);
  if (!row || row.blocked) return null;
  return row.id === openingRendition(rows)?.id ? null : row.id;
}

function size(width: number | null | undefined, height: number | null | undefined): PixelSize | null {
  return width && height && width > 0 && height > 0 ? { width, height } : null;
}

/**
 * An instance's ROW as a capture, for the sheet that looks at a day before
 * anything is fetched: the proxy it shows, the primary's own file, and the
 * companion Winnow paired with it — the same three files `materialize` turns
 * into a `MediaOrigin`, so the ids match what the workbench lists once the
 * picture crosses (`delivered:<name>` on both sides).
 *
 * `held` says which of them the session already holds, so the chip does not
 * promise a fetch that will not happen. Nothing is read from a head here: a
 * RAW's render size stays unmeasured until it is looked at.
 */
export function rowCaptureInput(
  row: WinnowAssetRow,
  host: string,
  held: (assetId: string) => boolean,
): CaptureInput {
  const assetId = `${host}/${row.id}`;
  const others = [
    {
      name: row.filename,
      bytes: row.file_size ?? null,
      here: held(assetId),
      assetId,
      ...(isRawImage(row.filename)
        ? { sensor: size(row.width, row.height) }
        : { pixels: size(row.width, row.height) }),
    },
  ];
  const cid = row.companion_id;
  const cname = row.companion_filename;
  const paired =
    cid && cname && (!row.group_kind || row.group_kind === 'raw_jpeg') && (!row.companion_media_type || row.companion_media_type === 'photo');
  if (paired) {
    const companionId = `${host}/${cid}`;
    others.push({
      name: cname,
      bytes: row.companion_file_size ?? null,
      here: held(companionId),
      assetId: companionId,
      ...(isRawImage(cname)
        ? { sensor: size(row.companion_width, row.companion_height) }
        : { pixels: size(row.companion_width, row.companion_height) }),
    });
  }
  return {
    open: { name: row.filename, bytes: null, here: true, assetId },
    openIsProxy: true,
    others,
  };
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Develop/CaptureView.swift
// The capture's files as a VIEWER lists them — the words on the lightbox's
// chips, what each one can show, and what the `Develop` verb under the picture
// carries away. Port of `src/shared/develop/capture-view.ts` (R6 of
// `docs/capture-renditions.md`).
//
// The viewer writes NOTHING — *"la visionneuse n'écrit aucun choix, c'est
// juste de la visualisation"*. Looking at the camera's JPEG beside the proxy
// is looking; what carries the choice is the VERB (`viewedRendition`). The
// fetching, the slicing of a RAW's render and the images are the app's.
//
// The web reads an instance's row as `WinnowAssetRow`; the kernel takes the
// few fields the viewer needs as `CaptureRow` (the wire's snake_case keys in
// lowerCamel), so a later port of the Winnow client maps its row onto it.

import Foundation

/// The rows a viewer can put on screen: everything but the SENSOR's. A viewer
/// draws files; developing a sensor plane is the Develop tool's work, and a
/// chip that can only say "open Develop" is a control that disappoints.
public func viewableRenditions(_ rows: [Rendition]) -> [Rendition] {
    rows.filter { $0.role != .sensor }
}

/// The chip's word — the fidelity chip's own: `Proxy`, else the file's name.
public func viewLabel(_ row: Rendition) -> String {
    row.role == .proxy ? "Proxy" : row.name
}

/// The line under the picture while that file is shown: what it is, its
/// pixels where measured, its weight, and whether a click will fetch it.
public func viewFacts(_ row: Rendition, _ format: (Int) -> String) -> String {
    var parts: [String] = []
    if row.role == .proxy {
        parts.append("the proxy, where the picture opens")
    } else if row.reach == .embedded {
        parts.append("the render the camera wrote inside the RAW")
    } else {
        parts.append("the file itself")
    }
    if let pixels = row.pixels { parts.append("\(pixels.width) × \(pixels.height)") }
    if let bytes = row.bytes { parts.append(format(bytes)) }
    if !row.here, let assetId = row.assetId, !assetId.isEmpty {
        parts.append("fetched from its instance on request, held for this session")
    }
    return parts.joined(separator: " · ")
}

/// What the `Develop` verb hands to the tool: the rendition being viewed, as
/// the roll stores it — nil for the one the picture opens on, which is what
/// the workbench stores for "chose nothing", so a look at the proxy never
/// writes a choice the pill would then draw as one.
public func viewedRendition(_ rows: [Rendition], _ viewing: String?) -> String? {
    guard let viewing, !viewing.isEmpty else { return nil }
    guard let row = rows.first(where: { $0.id == viewing }), row.blocked == nil else { return nil }
    return row.id == openingRendition(rows)?.id ? nil : row.id
}

/// The fields of an instance's asset row the viewer reads.
public struct CaptureRow: Equatable, Sendable {
    public var id: Int
    public var filename: String
    public var width: Int?
    public var height: Int?
    public var fileSize: Int?
    /// `raw_jpeg`, `live_photo`, … — how the instance paired the companion.
    public var groupKind: String?
    public var companionId: Int?
    public var companionFilename: String?
    public var companionFileSize: Int?
    /// `photo` or `video`.
    public var companionMediaType: String?
    public var companionWidth: Int?
    public var companionHeight: Int?

    public init(id: Int, filename: String, width: Int? = nil, height: Int? = nil, fileSize: Int? = nil,
                groupKind: String? = nil, companionId: Int? = nil, companionFilename: String? = nil,
                companionFileSize: Int? = nil, companionMediaType: String? = nil,
                companionWidth: Int? = nil, companionHeight: Int? = nil) {
        self.id = id; self.filename = filename; self.width = width; self.height = height; self.fileSize = fileSize
        self.groupKind = groupKind; self.companionId = companionId; self.companionFilename = companionFilename
        self.companionFileSize = companionFileSize; self.companionMediaType = companionMediaType
        self.companionWidth = companionWidth; self.companionHeight = companionHeight
    }
}

private func rowSize(_ width: Int?, _ height: Int?) -> PixelSize? {
    guard let width, let height, width > 0, height > 0 else { return nil }
    return PixelSize(width: width, height: height)
}

/// One of the row's files as a capture's file: a RAW's stated pixels are its SENSOR's.
private func rowFile(_ name: String, _ bytes: Int?, _ here: Bool, _ assetId: String, _ width: Int?, _ height: Int?) -> CaptureFile {
    var file = CaptureFile(name: name, bytes: bytes, here: here, assetId: assetId)
    if isRawImage(name) { file.sensor = rowSize(width, height) } else { file.pixels = rowSize(width, height) }
    return file
}

/// An instance's ROW as a capture, for the sheet that looks at a day before
/// anything is fetched: the proxy it shows, the primary's own file, and the
/// companion the instance paired with it — the same three files a fetched
/// picture's origin names, so the ids match what the workbench lists once the
/// picture crosses (`delivered:<name>` on both sides). `held` says which of
/// them the session already holds. Nothing is read from a head here: a RAW's
/// render size stays unmeasured until it is looked at. A companion that is not
/// a photograph — a Live Photo's `.mov` — is never a file of the capture.
public func rowCaptureInput(_ row: CaptureRow, _ host: String, _ held: (String) -> Bool) -> CaptureInput {
    let assetId = "\(host)/\(row.id)"
    var others = [rowFile(row.filename, row.fileSize, held(assetId), assetId, row.width, row.height)]
    if let cid = row.companionId, cid != 0, let cname = row.companionFilename, !cname.isEmpty,
       row.groupKind == nil || row.groupKind == "" || row.groupKind == "raw_jpeg",
       row.companionMediaType == nil || row.companionMediaType == "" || row.companionMediaType == "photo" {
        let companionId = "\(host)/\(cid)"
        others.append(rowFile(cname, row.companionFileSize, held(companionId), companionId, row.companionWidth, row.companionHeight))
    }
    return CaptureInput(
        open: CaptureFile(name: row.filename, bytes: nil, here: true, assetId: assetId),
        openIsProxy: true,
        others: others
    )
}
